x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheFile.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.searchablesnapshots.cache;
+
+import org.apache.lucene.store.AlreadyClosedException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.CheckedBiFunction;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.util.concurrent.AbstractRefCounted;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+class CacheFile implements Releasable {
+
+    @FunctionalInterface
+    public interface EvictionListener {
+        void onEviction(CacheFile evictedCacheFile);
+    }
+
+    private static final int RANGE_SIZE = 1 << 15;
+
+    private static final StandardOpenOption[] OPEN_OPTIONS = new StandardOpenOption[]{
+        StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.SPARSE
+    };
+
+    private final AbstractRefCounted refCounter = new AbstractRefCounted("CacheFile") {
+        @Override
+        protected void closeInternal() {
+            CacheFile.this.closeInternal();
+        }
+    };
+
+    private final SparseFileTracker tracker;
+    private final String name;
+    private final Path file;
+
+    private volatile Set<EvictionListener> listeners;
+    private volatile FileChannel channel;
+    private volatile boolean closed;
+
+    CacheFile(String name, long length, Path file) {
+        this.tracker = new SparseFileTracker(file.toString(), length);
+        this.name = Objects.requireNonNull(name);
+        this.file = Objects.requireNonNull(file);
+        this.listeners = new HashSet<>();
+        this.closed = false;
+        assert invariant();
+    }
+
+    public String getName() {
```

Comment:
I removed the method in 50f36bbcae3e1efe07c5e7f0b1180cdea1e19579.

> From my experience working with these things so far I think it would also be very useful to include the `ShardId` in the `toString()` output as well.

I agree. If that's OK I'd like to address this along with [another TODO](https://github.com/elastic/elasticsearch/pull/50693/files#diff-a0405cffc7e6de74715aa3935ac736ecR118) I have in a follow up PR.

Reply:
> From my experience working with these things so far I think it would also be very useful to include the `ShardId` in the `toString()` output as well.

This is addressed in https://github.com/elastic/elasticsearch/pull/51669